### ISSUE TITLE
"Assertion failed: ESI.Type != EST_Uninstantiated" with clang-cl 12.0.1 on Windows

### ISSUE THREAD

**Issue by jgreitemann** [llvm/llvm-project#54371]
### System Information
OS: Windows 10 Pro 21H1 (Build 19043.1586)
clang version 12.0.1 (https://github.com/llvm/llvm-project.git fed41342a82f5a3a9201819a82bf7a48313e296b)
Target: x86_64-pc-windows-msvc
Thread model: posix

### Crash backtrace:
```
$ ./Action-5f2ce2.sh
Assertion failed: ESI.Type != EST_Uninstantiated, file C:\Users\berglerm\Desktop\llvm-project\clang\lib\Sema\SemaTemplateInstantiate.cpp, line 2201
PLEASE submit a bug report to https://bugs.llvm.org/ and include the crash backtrace, preprocessed source, and associated run script.
Stack dump:
0.      Program arguments: S:\\dist\\clang-12.0.1-x64-win64\\bin\\clang-cl.exe -cc1 -triple x86_64-pc-windows-msvc19.29.30137 -fsyntax-only -disable-free -main-file-name Action.cpp -mrelocation-model pic -pic-level 2 -mframe-pointer=none -relaxed-aliasing -fmath-errno -fno-rounding-math -mconstructor-aliases -munwind-tables -target-cpu x86-64 -mllvm -x86-asm-syntax=intel -tune-cpu generic -D_DEBUG -D_MT -D_DLL --dependent-lib=msvcrtd --dependent-lib=oldnames -stack-protector 2 -fcxx-exceptions -fexceptions -fexternc-nounwind -fms-volatile -fdiagnostics-format msvc -gno-column-info -gcodeview -debug-info-kind=limited -D BOOST_THREAD_PROVIDES_FUTURE -D BOOST_THREAD_PROVIDES_FUTURE_CONTINUATION -D CEREAL_THREAD_SAFE=1 -D FMT_LOCALE -D JSON_USE_IMPLICIT_CONVERSIONS=1 -D NOMINMAX -D UNICODE -D WIN32 -D _WINDOWS -O0 -fdeprecated-macro -ferror-limit 19 -fno-use-cxa-atexit -fms-extensions -fms-compatibility -fms-compatibility-version=19.29.30137 -std=c++17 -fdelayed-template-parsing -fno-inline -faddrsig -x c++ Action-5f2ce2.cpp
1.      C:/Users/greitemannj/communicator/plugins/mqtt/MVJSON/Action.cpp:132:87: current parser token ')'
2.      C:/Users/greitemannj/communicator/plugins/mqtt/MVJSON/Action.cpp:56:1: parsing namespace 'MVJSON'
3.      C:/Users/greitemannj/communicator/plugins/mqtt/MVJSON/Action.cpp:109:1: parsing function body 'MVJSON::from_json'
4.      C:/Users/greitemannj/communicator/plugins/mqtt/MVJSON/Action.cpp:109:1: in compound statement ('{}')
5.      C:/Users/greitemannj/communicator/plugins/mqtt/MVJSON/Action.cpp:114:3: in compound statement ('{}')
6.      C:/Users/greitemannj/communicator/plugins/mqtt/MVJSON/Action.cpp:128:5: in compound statement ('{}')
 #0 0x00007ff6ea574f45 (S:\dist\clang-12.0.1-x64-win64\bin\clang-cl.exe+0x1ec4f45)
 #1 0x00007ffe88621881 (C:\WINDOWS\System32\ucrtbase.dll+0x71881)
 #2 0x00007ffe88622851 (C:\WINDOWS\System32\ucrtbase.dll+0x72851)
 #3 0x00007ffe8862426e (C:\WINDOWS\System32\ucrtbase.dll+0x7426e)
 #4 0x00007ffe88624165 (C:\WINDOWS\System32\ucrtbase.dll+0x74165)
 #5 0x00007ffe886244f1 (C:\WINDOWS\System32\ucrtbase.dll+0x744f1)
 #6 0x00007ff6ec08e45f (S:\dist\clang-12.0.1-x64-win64\bin\clang-cl.exe+0x39de45f)
 #7 0x00007ff6ec1bcc85 (S:\dist\clang-12.0.1-x64-win64\bin\clang-cl.exe+0x3b0cc85)
 #8 0x00007ff6ec1aa13f (S:\dist\clang-12.0.1-x64-win64\bin\clang-cl.exe+0x3afa13f)
 #9 0x00007ff6ebb8a1b4 (S:\dist\clang-12.0.1-x64-win64\bin\clang-cl.exe+0x34da1b4)
#10 0x00007ff6ec1b447a (S:\dist\clang-12.0.1-x64-win64\bin\clang-cl.exe+0x3b0447a)
#11 0x00007ff6ec1b34f4 (S:\dist\clang-12.0.1-x64-win64\bin\clang-cl.exe+0x3b034f4)
#12 0x00007ff6ec1748b4 (S:\dist\clang-12.0.1-x64-win64\bin\clang-cl.exe+0x3ac48b4)
#13 0x00007ff6ec1bb94e (S:\dist\clang-12.0.1-x64-win64\bin\clang-cl.exe+0x3b0b94e)
#14 0x00007ff6ec1aded9 (S:\dist\clang-12.0.1-x64-win64\bin\clang-cl.exe+0x3afded9)
#15 0x00007ff6ec1b7835 (S:\dist\clang-12.0.1-x64-win64\bin\clang-cl.exe+0x3b07835)
#16 0x00007ff6ec1b40dd (S:\dist\clang-12.0.1-x64-win64\bin\clang-cl.exe+0x3b040dd)
#17 0x00007ff6ec14fe27 (S:\dist\clang-12.0.1-x64-win64\bin\clang-cl.exe+0x3a9fe27)
#18 0x00007ff6ec14ec7f (S:\dist\clang-12.0.1-x64-win64\bin\clang-cl.exe+0x3a9ec7f)
#19 0x00007ff6ec17cfa5 (S:\dist\clang-12.0.1-x64-win64\bin\clang-cl.exe+0x3accfa5)
#20 0x00007ff6ec156275 (S:\dist\clang-12.0.1-x64-win64\bin\clang-cl.exe+0x3aa6275)
#21 0x00007ff6ebebda54 (S:\dist\clang-12.0.1-x64-win64\bin\clang-cl.exe+0x380da54)
#22 0x00007ff6ebeb1fe2 (S:\dist\clang-12.0.1-x64-win64\bin\clang-cl.exe+0x3801fe2)
#23 0x00007ff6ebb1fbbb (S:\dist\clang-12.0.1-x64-win64\bin\clang-cl.exe+0x346fbbb)
#24 0x00007ff6ebb1ad76 (S:\dist\clang-12.0.1-x64-win64\bin\clang-cl.exe+0x346ad76)
#25 0x00007ff6ebb19538 (S:\dist\clang-12.0.1-x64-win64\bin\clang-cl.exe+0x3469538)
#26 0x00007ff6ebb15b5e (S:\dist\clang-12.0.1-x64-win64\bin\clang-cl.exe+0x3465b5e)
#27 0x00007ff6ebb1c3f0 (S:\dist\clang-12.0.1-x64-win64\bin\clang-cl.exe+0x346c3f0)
#28 0x00007ff6ebb1f79e (S:\dist\clang-12.0.1-x64-win64\bin\clang-cl.exe+0x346f79e)
#29 0x00007ff6ebb1ad76 (S:\dist\clang-12.0.1-x64-win64\bin\clang-cl.exe+0x346ad76)
#30 0x00007ff6ebb15b5e (S:\dist\clang-12.0.1-x64-win64\bin\clang-cl.exe+0x3465b5e)
#31 0x00007ff6ebb1c31a (S:\dist\clang-12.0.1-x64-win64\bin\clang-cl.exe+0x346c31a)
#32 0x00007ff6ebb36de1 (S:\dist\clang-12.0.1-x64-win64\bin\clang-cl.exe+0x3486de1)
#33 0x00007ff6ebb3c440 (S:\dist\clang-12.0.1-x64-win64\bin\clang-cl.exe+0x348c440)
#34 0x00007ff6ebb3b758 (S:\dist\clang-12.0.1-x64-win64\bin\clang-cl.exe+0x348b758)
#35 0x00007ff6ebb35df0 (S:\dist\clang-12.0.1-x64-win64\bin\clang-cl.exe+0x3485df0)
#36 0x00007ff6ebb3bbe2 (S:\dist\clang-12.0.1-x64-win64\bin\clang-cl.exe+0x348bbe2)
#37 0x00007ff6ebb3b758 (S:\dist\clang-12.0.1-x64-win64\bin\clang-cl.exe+0x348b758)
#38 0x00007ff6ebb390ae (S:\dist\clang-12.0.1-x64-win64\bin\clang-cl.exe+0x34890ae)
#39 0x00007ff6ebb3bc2c (S:\dist\clang-12.0.1-x64-win64\bin\clang-cl.exe+0x348bc2c)
#40 0x00007ff6ebb3b758 (S:\dist\clang-12.0.1-x64-win64\bin\clang-cl.exe+0x348b758)
#41 0x00007ff6ebb35df0 (S:\dist\clang-12.0.1-x64-win64\bin\clang-cl.exe+0x3485df0)
#42 0x00007ff6ebb3bbe2 (S:\dist\clang-12.0.1-x64-win64\bin\clang-cl.exe+0x348bbe2)
#43 0x00007ff6ebb3b758 (S:\dist\clang-12.0.1-x64-win64\bin\clang-cl.exe+0x348b758)
#44 0x00007ff6ebb3b533 (S:\dist\clang-12.0.1-x64-win64\bin\clang-cl.exe+0x348b533)
#45 0x00007ff6ebb3c987 (S:\dist\clang-12.0.1-x64-win64\bin\clang-cl.exe+0x348c987)
#46 0x00007ff6ebb3bc40 (S:\dist\clang-12.0.1-x64-win64\bin\clang-cl.exe+0x348bc40)
#47 0x00007ff6ebb3b758 (S:\dist\clang-12.0.1-x64-win64\bin\clang-cl.exe+0x348b758)
#48 0x00007ff6ebb35df0 (S:\dist\clang-12.0.1-x64-win64\bin\clang-cl.exe+0x3485df0)
#49 0x00007ff6ebb38681 (S:\dist\clang-12.0.1-x64-win64\bin\clang-cl.exe+0x3488681)
#50 0x00007ff6ebabdd57 (S:\dist\clang-12.0.1-x64-win64\bin\clang-cl.exe+0x340dd57)
#51 0x00007ff6ebadd39d (S:\dist\clang-12.0.1-x64-win64\bin\clang-cl.exe+0x342d39d)
#52 0x00007ff6ebabbcda (S:\dist\clang-12.0.1-x64-win64\bin\clang-cl.exe+0x340bcda)
#53 0x00007ff6ebabcc5c (S:\dist\clang-12.0.1-x64-win64\bin\clang-cl.exe+0x340cc5c)
#54 0x00007ff6ebb4d99d (S:\dist\clang-12.0.1-x64-win64\bin\clang-cl.exe+0x349d99d)
#55 0x00007ff6ebb5136e (S:\dist\clang-12.0.1-x64-win64\bin\clang-cl.exe+0x34a136e)
#56 0x00007ff6ebade2fd (S:\dist\clang-12.0.1-x64-win64\bin\clang-cl.exe+0x342e2fd)
#57 0x00007ff6ebabc5d1 (S:\dist\clang-12.0.1-x64-win64\bin\clang-cl.exe+0x340c5d1)
#58 0x00007ff6ebac0e0b (S:\dist\clang-12.0.1-x64-win64\bin\clang-cl.exe+0x3410e0b)
#59 0x00007ff6ebab7722 (S:\dist\clang-12.0.1-x64-win64\bin\clang-cl.exe+0x3407722)
#60 0x00007ff6eade6e96 (S:\dist\clang-12.0.1-x64-win64\bin\clang-cl.exe+0x2736e96)
#61 0x00007ff6eade6ca8 (S:\dist\clang-12.0.1-x64-win64\bin\clang-cl.exe+0x2736ca8)
#62 0x00007ff6eadaf744 (S:\dist\clang-12.0.1-x64-win64\bin\clang-cl.exe+0x26ff744)
#63 0x00007ff6eae5ebd2 (S:\dist\clang-12.0.1-x64-win64\bin\clang-cl.exe+0x27aebd2)
#64 0x00007ff6e875ee5b (S:\dist\clang-12.0.1-x64-win64\bin\clang-cl.exe+0xaee5b)
#65 0x00007ff6e8759be9 (S:\dist\clang-12.0.1-x64-win64\bin\clang-cl.exe+0xa9be9)
#66 0x00007ff6e875bb75 (S:\dist\clang-12.0.1-x64-win64\bin\clang-cl.exe+0xabb75)
#67 0x00007ff6ec8b91c0 (S:\dist\clang-12.0.1-x64-win64\bin\clang-cl.exe+0x42091c0)
#68 0x00007ffe894b7034 (C:\WINDOWS\System32\KERNEL32.DLL+0x17034)
#69 0x00007ffe8ad22651 (C:\WINDOWS\SYSTEM32\ntdll.dll+0x52651)
```

### Preprocessed source & run script:
[Action-5f2ce2.zip](https://github.com/llvm/llvm-project/files/8245102/Action-5f2ce2.zip)

**Comment by EugeneZelenko** [comment]
Does same crash happen in 13, 14 or `main`?

**Comment by jgreitemann** [comment]
The crash is fixed as of 13.0.0.
